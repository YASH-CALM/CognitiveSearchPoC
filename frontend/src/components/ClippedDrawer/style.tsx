/**
 * 
 * @param open if true width is 95% otherwise width is 100%
 */
export const documentsStyle = (open: boolean): React.CSSProperties => ({
    marginLeft: 'auto',
    marginRight: 'auto',
    width: open ? '95%' : '100%',
})

export const nav: React.CSSProperties = {
    height: '100vh',
    width: '100%',
}

/**
 * 
 * @param theme the react-uwp theme
 */
export const searchBarStyle = (theme: any): React.CSSProperties => ({
    background: theme.acrylicTexture40.background,
    color: '#EEEEEE88',
    display: 'block',
    margin: '0 auto',
    width: '90%'
})

export const checkBoxContainerStyle: React.CSSProperties = {
    margin: '0 auto',
    width: '80%'
}

/**
 * 
 * @param open if true hamburger icon is hidden otherwise hamburger icon is visible
 */
export const hamburg = (open: boolean): React.CSSProperties => ({
    left: 0,
    position: 'absolute',
    top: 0,
    visibility: open ? 'hidden' : 'visible',
})

/**
 * 
 * @param open if open uses navWidthOpen and mainWidthOpen values otherwise uses navWidthClose and mainWidthClose values
 * @param navWidthOpen position from left when open or closed
 * @param mainWidthOpen width when open or closed
 */
export const titleStyle = (open: boolean, navWidth: {[key: string]: string | number}, mainWidth: {[key: string]: string | number}): React.CSSProperties => ({
    borderTop: '2px solid #88888888',
    bottom: 4,
    height: 36,
    left: open ? navWidth.open : navWidth.close,
    position: 'absolute',
    transition: "left .25s ease-in-out, width .25s ease-in-out",
    width: open ? mainWidth.open : mainWidth.close,
})

