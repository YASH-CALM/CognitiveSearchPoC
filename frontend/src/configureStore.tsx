import { applyMiddleware, compose, createStore } from 'redux';
import thunkMiddleWare from 'redux-thunk';
import rootReducer from './reducers';

const DEVTOOLS = '__REDUX_DEVTOOLS_EXTENSION__';
const enhancer = window[DEVTOOLS] ? window[DEVTOOLS]() : (f: any) => f;

/**
 * creates the store from the rootReducer, middleware, and redux devtools
 */
export default () => (createStore(rootReducer,
  compose(
    applyMiddleware(thunkMiddleWare),
    enhancer
  ))
);