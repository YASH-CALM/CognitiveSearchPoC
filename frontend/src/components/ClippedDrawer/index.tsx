import * as PropTypes from 'prop-types';
import * as React from 'react';
import CheckBox from 'react-uwp/CheckBox';
import IconButton from 'react-uwp/IconButton';
import NavigationView from 'react-uwp/NavigationView';
import { IDocument, IPage } from '../../types/documents';
import { IGraph } from '../../types/graph';
import { IInfoboxWithPushPin } from '../../types/pins';
import Documents from '../Documents';
import ForceDirectedGraph from '../ForceDirectedGraph';
import GraphIcon from '../GraphIcon';
import Main from '../Main';
import Map from '../Map';
import SearchBox from '../SearchBox';
import Title from '../Title';
import * as Style from './style';

export interface IChecked<T> {
    checked: boolean,
    value: T,
}

export interface IClippedDrawerProps extends React.HTMLAttributes<HTMLDivElement>{
    children?: object,
    display: string,
    docs: IDocument[],
    graph: IGraph,
    onSearchClick: (query: string) => void,
    onViewClick: (display: string) => void,
    pins: IInfoboxWithPushPin[],
    query: string,
}

interface IClippedDrawerState {
    open: boolean,
    search: string,
    tags: Array<IChecked<string>>,
}

const docsToUniqueTags = (docs: IDocument[]): Array<IChecked<string>> => {
    const o: object = {};
    docs.reduce((a: IPage[], i: IDocument) => [...a, ...i.pages], []).reduce((a: string[], i: IPage) => [...a, ...i.tags], []).forEach((t: string) => o[t] = 0);
    return Object.keys(o).map(t => ({checked: false, value: t}));
}

class ClippedDrawer extends React.Component<IClippedDrawerProps,IClippedDrawerState> {
    public static contextTypes = {theme: PropTypes.object};

    constructor(props: IClippedDrawerProps) {
        super(props);
        this.state = {
            open: true,
            search: props.query,
            tags: docsToUniqueTags(props.docs),
        }
    }

    public componentWillReceiveProps(nextProps: IClippedDrawerProps) {
        if(JSON.stringify(this.props.docs) !== JSON.stringify(nextProps.docs)) {
            const tags = docsToUniqueTags(nextProps.docs);
            this.setState({tags});
        }
    }

    public onNavClick = (e: any) => {
        const open = !this.state.open;
        this.setState({open});
    }

    public onSearchChange = (e: any) => {
        this.setState({search: e.target.value});
    }

    public onCheckBoxClick = (id: number) => (e: any) => {
        const tags = [...this.state.tags];
        tags[id].checked = !tags[id].checked;
        this.setState({tags})
    }

    public render() {
    const { display, docs, graph, onViewClick, pins } = this.props;
    const mainWidth: {[key: string]: number | string} = {
        close: '100%',
        open: window.innerWidth > 500 ? '67%' : '0%',
    }
    const navWidth: {[key: string]: number | string} = {
        close: 0,
        open: window.innerWidth > 500 ? '33%' : '100%',
    }
    const { theme } = this.context;
    const documentsStyle: React.CSSProperties = Style.documentsStyle(this.state.open);
    const nav: React.CSSProperties = Style.nav;
    const checkBoxContainerStyle: React.CSSProperties = Style.checkBoxContainerStyle;
    const hamburg: React.CSSProperties = Style.hamburg(this.state.open);
    const titleStyle: React.CSSProperties = Style.titleStyle(this.state.open, navWidth, mainWidth);

    const topIcon = <IconButton onClick={this.onNavClick}>GlobalNavButton</IconButton>
    const tags = (this.state.open ? this.state.tags.map((t,i) => 
    <div key={t.value} style={checkBoxContainerStyle}>
    <CheckBox
        id={`${t}--${i}`}
        defaultChecked={t.checked}
        onClick={this.onCheckBoxClick(i)}
        label={t.value}
        />
    </div>) : []);
    const navTopNodes = [
        <SearchBox
          key={1}
          value={this.state.search}
          onChange={this.onSearchChange}
          placeholder={'Search'}
          searchAction={this.props.onSearchClick}
        />,
        ...tags
    ]
    const onClick = (d: string) => (e: any) => onViewClick(d);
    const visibleDocs: IDocument[] = this.state.tags.reduce((a, i) => a || i.checked, false) ?
    docs.filter(doc => {
        const docTags = doc.pages.reduce((a, i) => [...a, ...i.tags], []);
        return this.state.tags.reduce((a, i) => a || (i.checked && docTags.indexOf(i.value) !== -1), false)
    }
    ) : docs;
    const onWidthResize = (width: number): number => width*(this.state.open ? Number((mainWidth.open as string).slice(0,-1))/100 : 1);
    const onHeightResize = (height: number): number => height-92;
    return (
        <>
            <NavigationView
            isControlled={true}
            displayMode={'minimal'}
            topIcon={topIcon}
            autoResize={false}
            defaultExpanded={true}
            expandedWidth={this.state.open ? navWidth.open : navWidth.close}
            style={{...nav, background: theme.acrylicTexture40.background}}
            pageTitle={''}
            focusNavigationNodeIndex={2}
            navigationTopNodes={navTopNodes}
            >
            <IconButton onClick={this.onNavClick} style={hamburg}>
                GlobalNavButton
            </IconButton>
            <Title style={titleStyle}/>
            {
                (window.innerWidth > 500 || !this.state.open && window.innerWidth <= 500) &&
                <Main
                display={display}
                onClick={onClick}
                open={this.state.open}
                views={[
                    {component: <Documents key={'Results'} docs={visibleDocs} pageSize={10} pagePadding={window.innerWidth > 320 ? 2 : 1} pagenationWidth={this.state.open ? mainWidth.open : mainWidth.close} style={documentsStyle}/>, display: 'Results', icon: 'List'},
                    {component: <Map key={'Maps'} infoboxesWithPushPins={pins}/>, display: 'Maps', icon: 'MapPinLegacy'},
                    {component: <ForceDirectedGraph key={'Graph'} graph={graph} width={onWidthResize} height={onHeightResize}/>, display: 'Graph', icon: <GraphIcon/>},
                ]}
                />
            }   
            </NavigationView>
            </>
  );
}}

export default ClippedDrawer;