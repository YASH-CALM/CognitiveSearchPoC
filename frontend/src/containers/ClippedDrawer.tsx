import { connect } from 'react-redux';
import { IDocument } from 'src/types/documents';
import { IGraph } from 'src/types/graph';
import { IInfoboxWithPushPin } from 'src/types/pins';
import { setDisplay } from '../actions/display';
import { fetchQuery } from '../actions/query';
import ClippedDrawer from '../components/ClippedDrawer';
import { IState } from '../types';

interface IOwnProps {
  children?: any,
}

interface IPropsFromState extends IOwnProps {
  display: string,
  docs: IDocument[],
  graph: IGraph,
  pins: IInfoboxWithPushPin[],
  query: string,
}

interface IPropsFromDispatch {
  onSearchClick: typeof fetchQuery,
  onViewClick: typeof setDisplay,
}

const mapStateToProps = (state: IState, ownProps: IOwnProps): IPropsFromState => ({
  display: state.display,
  docs: state.documents,
  graph: state.graph,
  pins: state.pins,
  query: state.query,
  ...ownProps
})

const mapDispatchToProps: IPropsFromDispatch = ({
  onSearchClick: fetchQuery,
  onViewClick: setDisplay,
})

export default connect(mapStateToProps, mapDispatchToProps)(ClippedDrawer);