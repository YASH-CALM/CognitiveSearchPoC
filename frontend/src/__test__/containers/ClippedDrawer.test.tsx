import * as React from 'react';
import { Provider } from 'react-redux';
import { createRenderer as ShallowRenderer } from 'react-test-renderer/shallow';
import configureStore from '../../configureStore';
import ClippedDrawer from '../../containers/ClippedDrawer';

jest.mock('../../util/docs', () => ({
    docs: jest.fn(() => [])
}))

const store = configureStore();

describe('ClippedDrawer Container', () => {
    const renderer = ShallowRenderer()
    it('should match snapshot', () => {
        renderer.render(<Provider store={store}>
            <ClippedDrawer/>
            </Provider>
            )
        const result = renderer.getRenderOutput()
        expect(result).toMatchSnapshot()
    })
})