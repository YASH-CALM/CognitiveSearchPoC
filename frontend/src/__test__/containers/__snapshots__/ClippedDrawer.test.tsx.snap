// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClippedDrawer Container should match snapshot 1`] = `
<Context.Provider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "storeState": Object {
        "display": "Results",
        "documents": Array [],
        "graph": Object {
          "links": Array [
            Object {
              "source": "A",
              "target": "B",
              "value": 1,
            },
            Object {
              "source": "A",
              "target": "C",
              "value": 1,
            },
            Object {
              "source": "A",
              "target": "D",
              "value": 1,
            },
            Object {
              "source": "A",
              "target": "E",
              "value": 1,
            },
            Object {
              "source": "F",
              "target": "A",
              "value": 1,
            },
            Object {
              "source": "G",
              "target": "A",
              "value": 1,
            },
            Object {
              "source": "H",
              "target": "A",
              "value": 1,
            },
            Object {
              "source": "I",
              "target": "A",
              "value": 1,
            },
            Object {
              "source": "J",
              "target": "A",
              "value": 1,
            },
            Object {
              "source": "K",
              "target": "A",
              "value": 1,
            },
            Object {
              "source": "L",
              "target": "A",
              "value": 1,
            },
            Object {
              "source": "M",
              "target": "A",
              "value": 1,
            },
            Object {
              "source": "B",
              "target": "N",
              "value": 10,
            },
            Object {
              "source": "C",
              "target": "O",
              "value": 10,
            },
            Object {
              "source": "D",
              "target": "P",
              "value": 10,
            },
            Object {
              "source": "E",
              "target": "Q",
              "value": 10,
            },
            Object {
              "source": "F",
              "target": "R",
              "value": 10,
            },
            Object {
              "source": "G",
              "target": "S",
              "value": 10,
            },
            Object {
              "source": "H",
              "target": "T",
              "value": 10,
            },
            Object {
              "source": "I",
              "target": "U",
              "value": 10,
            },
            Object {
              "source": "J",
              "target": "V",
              "value": 10,
            },
            Object {
              "source": "K",
              "target": "W",
              "value": 10,
            },
            Object {
              "source": "L",
              "target": "X",
              "value": 10,
            },
            Object {
              "source": "M",
              "target": "Y",
              "value": 10,
            },
            Object {
              "source": "M",
              "target": "Z",
              "value": 10,
            },
          ],
          "nodes": Array [
            Object {
              "group": 1,
              "id": "A",
            },
            Object {
              "group": 2,
              "id": "B",
            },
            Object {
              "group": 2,
              "id": "C",
            },
            Object {
              "group": 2,
              "id": "D",
            },
            Object {
              "group": 2,
              "id": "E",
            },
            Object {
              "group": 2,
              "id": "F",
            },
            Object {
              "group": 2,
              "id": "G",
            },
            Object {
              "group": 2,
              "id": "H",
            },
            Object {
              "group": 2,
              "id": "I",
            },
            Object {
              "group": 2,
              "id": "J",
            },
            Object {
              "group": 2,
              "id": "K",
            },
            Object {
              "group": 2,
              "id": "L",
            },
            Object {
              "group": 2,
              "id": "M",
            },
            Object {
              "group": 3,
              "id": "N",
            },
            Object {
              "group": 3,
              "id": "O",
            },
            Object {
              "group": 3,
              "id": "P",
            },
            Object {
              "group": 3,
              "id": "Q",
            },
            Object {
              "group": 3,
              "id": "R",
            },
            Object {
              "group": 3,
              "id": "S",
            },
            Object {
              "group": 3,
              "id": "T",
            },
            Object {
              "group": 3,
              "id": "U",
            },
            Object {
              "group": 3,
              "id": "V",
            },
            Object {
              "group": 3,
              "id": "W",
            },
            Object {
              "group": 3,
              "id": "X",
            },
            Object {
              "group": 3,
              "id": "Y",
            },
            Object {
              "group": 3,
              "id": "Z",
            },
          ],
        },
        "pins": Array [
          Object {
            "infobox": Object {
              "description": "this location was entered as a longitude latitude pair \\"[36, -112]\\"",
              "title": "Grand Canyon",
            },
            "location": Array [
              36,
              -112,
            ],
            "pushPin": Object {
              "description": "Grand Canyon",
              "options": Object {
                "color": "orange",
              },
              "title": "Grand Canyon",
            },
          },
          Object {
            "infobox": Object {
              "description": "this location was entered as an address \\"841 Avenue of the Cities\\"",
              "title": "YASH Office",
            },
            "location": Array [
              "841 Avenue of the Cities",
            ],
            "pushPin": Object {
              "description": "Yash Office",
              "options": Object {
                "color": "#005ba1",
              },
              "title": "YASH",
            },
          },
          Object {
            "infobox": Object {
              "description": "this location was entered as a longitude latitude pair \\"[45, -84]\\"",
              "title": "Lake Superior",
            },
            "location": Array [
              45,
              -84,
            ],
            "pushPin": Object {
              "description": "Lake Superior",
              "options": Object {
                "color": "lightgreen",
              },
              "title": "Lake Superior",
            },
          },
          Object {
            "infobox": Object {
              "description": "this location was entered as the name of the location \\"Statue of Liberty\\"",
              "title": "Statue of Liberty",
            },
            "location": Array [
              "Statue of Liberty",
            ],
            "pushPin": Object {
              "description": "Statue of Liberty",
              "options": Object {
                "color": "grey",
              },
              "title": "Statue of Liberty",
            },
          },
        ],
        "query": "",
      },
    }
  }
>
  <Connect(ClippedDrawer) />
</Context.Provider>
`;
